// @nteract/transform-dataresource/lib/charts/grid.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DataResourceTransformGrid = void 0;

var _style = _interopRequireDefault(require("styled-jsx/style"));

var React = _interopRequireWildcard(require("react"));

var _reactTable = _interopRequireDefault(require("react-table"));

var _reactTable2 = _interopRequireDefault(require("../css/react-table"));

var _reactTableHocFixedColumns = _interopRequireDefault(require("react-table-hoc-fixed-columns"));

var ReactTableFixedColumns = (0, _reactTableHocFixedColumns.default)(_reactTable.default);

var DataResourceTransformGrid = function DataResourceTransformGrid(_ref) {
  var _ref$data = _ref.data,
      data = _ref$data.data,
      schema = _ref$data.schema,
      height = _ref.height;
  var tableColumns = schema.fields.map(function (f, i) {
    return {
      Header: f.name,
      accessor: f.name,
      fixed: schema.primaryKey.indexOf(f.name) !== -1 && "left"
    };
  });
  return React.createElement("div", {
    style: {
      width: "calc(100vw - 150px)"
    },
    className: "jsx-".concat(_reactTable2.default.__hash)
  }, React.createElement(ReactTableFixedColumns, {
    data: data,
    columns: tableColumns,
    style: {
      height: "".concat(height, "px")
    },
    className: "-striped -highlight"
  }), React.createElement(_style.default, {
    styleId: _reactTable2.default.__hash,
    css: _reactTable2.default
  }));
};

exports.DataResourceTransformGrid = DataResourceTransformGrid;
