// @nteract/transform-dataresource/lib/HTMLLegend.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime-corejs2/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime-corejs2/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime-corejs2/helpers/toConsumableArray"));

var React = _interopRequireWildcard(require("react"));

var _PalettePicker = _interopRequireDefault(require("./PalettePicker"));

var HTMLLegend = function HTMLLegend(_ref) {
  var values = _ref.values,
      colorHash = _ref.colorHash,
      valueHash = _ref.valueHash,
      _ref$colors = _ref.colors,
      colors = _ref$colors === void 0 ? [] : _ref$colors,
      setColor = _ref.setColor;
  return React.createElement("div", {
    style: {
      marginTop: "20px"
    }
  }, (values.length > 18 ? (0, _toConsumableArray2.default)(values.filter(function (d, i) {
    return i < 18;
  })).concat(["Other"]) : values).map(function (v, vi) {
    return colorHash[v] && React.createElement("span", {
      style: {
        display: "inline-block",
        minWidth: "80px",
        margin: "5px"
      },
      key: "legend-item-".concat(vi)
    }, React.createElement("span", {
      style: {
        display: "inline-block",
        width: "20px",
        height: "20px",
        marginRight: "5px",
        borderRadius: "20px",
        marginBottom: "-5px",
        background: colorHash[v]
      }
    }), v, valueHash[v] && valueHash[v] > 1 && "(".concat(valueHash[v], ")") || "");
  }), setColor && React.createElement(_PalettePicker.default, {
    colors: colors,
    updateColor: function updateColor(newColorArray) {
      setColor(newColorArray);
    }
  }));
};

var _default = HTMLLegend;
exports.default = _default;
