// platform-folders/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.getHomeFolder = getHomeFolder;
exports.getCacheFolder = getCacheFolder;
exports.getConfigHome = getConfigHome;
exports.getDataHome = getDataHome;
exports.getDesktopFolder = getDesktopFolder;
exports.getDocumentsFolder = getDocumentsFolder;
exports.getDownloadsFolder = getDownloadsFolder;
exports.getPicturesFolder = getPicturesFolder;
exports.getVideosFolder = getVideosFolder;
exports.getMusicFolder = getMusicFolder;
exports.getSaveGamesFolder = getSaveGamesFolder;
exports.default = getPath;
var folders = require('../build/Release/platform-folders');

function getHomeFolder() {
    return folders.getHomeFolder();
}

function getCacheFolder() {
    return folders.getCacheDir();
}

function getConfigHome() {
    return folders.getConfigHome();
}

function getDataHome() {
    return folders.getDataHome();
}

function getDesktopFolder() {
    return folders.getDesktopFolder();
}

function getDocumentsFolder() {
    return folders.getDocumentsFolder();
}

function getDownloadsFolder() {
    return folders.getDownloadsFolder();
}

function getPicturesFolder() {
    return folders.getPicturesFolder();
}

function getVideosFolder() {
    return folders.getVideosFolder();
}

function getMusicFolder() {
    return folders.getMusicFolder();
}

function getSaveGamesFolder() {
    return folders.getSaveGamesFolder();
}

function getPath(name) {
    switch (name) {
        case "home":
            return getHomeFolder();
        case "appData":
        case "appdata":
            return getDataHome();
        case "userData":
            return getConfigHome();
        case "desktop":
            return getDesktopFolder();
        case "documents":
            return getDocumentsFolder();
        case "downloads":
            return getDownloadsFolder();
        case "music":
            return getMusicFolder();
        case "pictures":
            return getPicturesFolder();
        case "videos":
            return getVideosFolder();
        case "cache":
            return getCacheFolder();
        case "savegames":
            return getSaveGamesFolder();
        default:
            return undefined;

    }
}
